src/TreeT.hs

```diff
@@ -0,0 +1,207 @@
+    -- splitting choices
+    go k (SplitT choices) s = do
+      branches <- forM choices \(tg, nxt) -> Branch tg <$> go k nxt s
+      pure (Split s branches)
+    -- the fmap (for performance win, since we can do better than the liftM from Control.Monad)
+    go k (FmapT g x) s = go (\s' x' -> k s' (g x')) x s
+    -- the ap (for performance win, since we can do better than the ap from Control.Monad)
+    go k (ApT fs xs) s =
+      go (\s1 f -> go (\s2 x -> k s2 (f x)) xs s1) fs s
+    -- the bind
+    go k (BindT this nxt) s =
+      go (\s1 x -> go k (nxt x) s1) this s
+
+-- | Specialized version of the runner for the 'Identity' monad.
+runTreeI :: forall tag state a. 
+            TreeT tag state Identity a -> state -> Tree tag state a
+runTreeI = go Done
+  where
+    go :: forall b r. 
+          (state -> b -> Tree tag state r)  -- continuation
+       -> TreeT tag state Identity b -> state
+       -> Tree tag state r
+    -- the simple ones
+    go _ EmptyT _ = Stop
+    go k (LiftT (Identity action)) s = k s action
+    go k GetT s = k s s
+    go k (PutT newS) _ = k newS ()
+    -- splitting choices
+    go k (SplitT choices) s =
+      let branches = flip map choices \(tg, nxt) -> Branch tg (go k nxt s)
+      in Split s branches
+    -- the fmap (for performance win, since we can do better than the liftM from Control.Monad)
+    go k (FmapT g x) s = 
+      go (\s' x' -> k s' (g x')) x s
+    -- the ap (for performance win, since we can do better than the ap from Control.Monad)
+    go k (ApT fs xs) s =
+      go (\s1 f -> go (\s2 x -> k s2 (f x)) xs s1) fs s
+    -- the bind
+    go k (BindT this nxt) s = 
+      go (\s1 x -> go k (nxt x) s1) this s
+
+-- | Specialized version to check a property.
+runTreeAny :: forall tag state f a r.
+              Alternative f =>
+              (a -> f r) ->
+              TreeT tag state Identity a -> state -> f r
+runTreeAny predicate = go (const predicate)
+  where
+    go :: (state -> b -> f q)
+       -> TreeT tag state Identity b
+       -> state -> f q
+    go _ EmptyT _ = empty
+    go p (LiftT (Identity action)) s = p s action
+    go p GetT s = p s s
+    go p (PutT newS) _ = p newS ()
+    go p (SplitT choices) s =
+      asum $ flip map choices \(_tg, nxt) -> go p nxt s
+    go p (FmapT g x) s = 
+      go (\s' x' -> p s' (g x')) x s
+    go p (ApT fs xs) s =
+      go (\s1 f -> go (\s2 x -> p s2 (f x)) xs s1) fs s
+    go p (BindT this nxt) s = 
+      go (\s1 x -> go p (nxt x) s1) this s
+
+{-# SPECIALIZE runTreeAny :: forall tag state a r.
+                             (a -> Maybe r) ->
+                             TreeT tag state Identity a -> state -> Maybe r #-}
+
+-- | Defines a class of monad which support pruning and branching.
+class Monad m => MonadTree tag m | m -> tag where
+  stop :: m a
+  branch :: [(tag, m a)] -> m a
+
+-- | The monad which defines actions which keep a 'state',
+-- and also may prune and branch.
+data TreeT tag state m a where
```

Comment:
This is much more than just lifting `Tree` to a transformer. Also, one of the purposes of this exercise was to explicitely detach the notion of state from the symbolic engine. With the `GetT` and `PutT` in here, we just hid that notion of state deeper. 

I tend to think it might be simpler to just make the one function that needed state return it explicitely. I haven't looked at it in detail though.

Reply:
I think that actually the notion of state in inherent in the engine, in all the constraints and information we are gathering. Maybe instead of `GetT` and `PutT` are not the best choices, and maybe `AddConstraintT` is the right move here, in the same way that `BranchT` has become a constructor in `TreeT`... 🤔 What do you think?